php/index.php

```diff
@@ -1,45 +1,204 @@
+*/
+function process_incoming_requests($incoming_request_uri, array $options) {
+  switch ($incoming_request_uri) {
+    case preg_match('/oauth/', $a):
+      try {
+        $me = process_client_redirect($_GET['code'], $options);
+        echo("<p>Here's some information about the user:</p>");
+        echo("<pre>");
+        print_r ($me);
+        echo("</pre>");
+        break;
+      } catch (Exception $e) {
+        echo("<p>Something exceptional happened while interacting with Clever.");
+        echo("<pre>");
+        print_r ($e);
+        echo("</pre>");
+      }
+    default:
+      // Our home page route will create a Clever Instant Login button for users from the district our $district_id is set to.
+      $sign_in_link = generate_sign_in_with_clever_link($options);
+      echo("<h1>Login!</h1>");
+      echo('<p>' . $sign_in_link . '</p>');
+      break;
+  }
+}
+
+/**
+* Processes incoming requests to our $client_redirect
+* 1. Exchanges incoming code parameter for a bearer token
+* 2. Uses bearer token in a request to Clever's "/me" API resource
+* @param   string $code         OAuth 2.0 exchange code received when our OAuth redirect was triggered
+* @param   array  $options      Options used for Clever API requests
+* @return  array  $me_response  Hash of Clever's response when identifying a bearer token's owner
+*/
+function process_client_redirect($code, array $options) {
+  $bearer_token = exchange_code_for_bearer_token($code, $options);
+  $request_options = array('method' => 'GET', 'bearer_token' => $bearer_token);
+  $me_response = retrieve_me_response_for_bearer_token($bearer_token, $options);
+
+  // Real world applications would store the bearer token and relevant information about the user at this stage.
+  return $me_response;
+}
+
+/**
+* Exchanges a $code value received in a $client_redirect for a bearer token
+* @param   string $code          OAuth 2.0 exchange code received when our OAuth redirect was triggered
+* @param   array  $options       Options used for Clever API requests
+* @throws  Exception if the bearer token cannot be retrieved
+* @return  string $bearer_token  The string value of a user's OAuth 2.0 access token
+*/
+function exchange_code_for_bearer_token($code, array $options) {
+  $data = array('code' => $code,
+                'grant_type' => 'authorization_code',
+                'redirect_uri' => $options['client_redirect_url']);
+  $request_options = array('method' => 'POST', 'data' => $data);
+  $response = request_from_clever($options['clever_oauth_tokens_url'], $request_options, $options);
+  // Evaluate if the response is successful
+  if ($response && $response['response_code'] && $response['response_code'] == 200) {
+    $bearer_token = $response['response']['access_token'];
+    return $bearer_token;
+  } else {
+    // Handle condition when $code cannot be exchanged for bearer token from Clever
+    throw new Exception("Cannot retrieve bearer token.");
```

Comment:
Because we're trying to teach developers to Do The Right Thing™ , we probably should handle a few of the interesting HTTP status codes and expose any related info (eg: Retry-After header for 503? ) 

Particularly:
* 301, 302 redirects
* 400, 401, 403, 404, 429 to inform various user errors
* 500, 503